src/lib/operators/aggregate_sort.cpp

```diff
@@ -0,0 +1,600 @@
+    resolve_data_type(data_type, [&](auto type) {
+      using ColumnDataType = typename decltype(type)::type;
+      std::optional<ColumnDataType> previous_value;
+      ChunkID chunk_id{0};
+      for (const auto& chunk : chunks) {
+        auto segment = chunk->get_segment(column_id);
+        segment_iterate<ColumnDataType>(*segment, [&](const auto& position) {
+          if (previous_value && position.value() != *previous_value) {
+            group_boundaries.insert(RowID{chunk_id, position.chunk_offset()});
+          }
+          // TODO(anybody) shouldnt it be sufficient to do this in an else-block?
+          // It should work both ways and as far as I know both ways are used in Hyrise
+          // With one way we can save the emplace and with the other way we save the branching.
+          previous_value.emplace(position.value());
+        });
+        chunk_id++;
+      }
+    });
+  }
+
+  /*
+   * For all group by columns
+   *   For each group by column
+   *     output the column value at the start of the group (it is per definition the same in the whole group)
+   * Write outputted values into the result table
+   */
+  size_t groupby_index = 0;
+  for (const auto& column_id : _groupby_column_ids) {
+    auto group_boundary_iter = group_boundaries.cbegin();
+    auto data_type = input_table->column_data_type(column_id);
+    resolve_data_type(data_type, [&](auto type) {
+      using ColumnDataType = typename decltype(type)::type;
+      auto values = std::vector<ColumnDataType>(group_boundaries.size() + 1);
+      auto null_values = std::vector<bool>(group_boundaries.size() + 1);
+      size_t value_index = 0;
+      ChunkID chunk_id{0};
+      for (const auto& chunk : chunks) {
+        auto segment = chunk->get_segment(column_id);
+        segment_iterate<ColumnDataType>(*segment, [&](const auto& position) {
+          // ToDo access the row ids directly without iterating
+          // thoughts: its just a small part, but this should reduce runtime for this loop from O(input) to O(output)
+          const auto current_row_id = RowID{chunk_id, position.chunk_offset()};
+          const auto is_new_group = value_index == 0 || current_row_id == *group_boundary_iter;
+          if (is_new_group) {
+            null_values[value_index] = position.is_null();
+            values[value_index] = position.value();
+            if (value_index != 0) {
+              group_boundary_iter++;
+            }
+            value_index++;
+          }
+        });
+        chunk_id++;
+      }
+      // Write group by segments
+      _output_segments[groupby_index] =
+          std::make_shared<ValueSegment<ColumnDataType>>(std::move(values), std::move(null_values));
+    });
+    groupby_index++;
+  }
+
+  // Call _aggregate_values for each aggregate
+  uint64_t aggregate_index = 0;
+  for (const auto& aggregate : _aggregates) {
+    /*
+     * Special case for COUNT(*), which is the only case where !aggregate.column is true:
```

Comment:
Is this the same comment as above?

Reply:
Almost, only the variable name changed. However, "above" is 150 lines higher. Would you prefer a reference to the other comment?

Reply:
I don't have a strong opinion here.

Reply:
We'd keep it the way it is now, then. For us, it wasn't obvious where to find the other comment.
If anyone has a strong opinion on this: I'll keep the comment open one more day.